bloom-housing/bloom
# feat: new checkbox behavior

# Pull Request Template

## Issue

Addresses #1229 

## Description

The new OHA preference set has new "exclusive" options, meaning options that when checked should un-check all other boxes, and if checked must be the only checked box.

This logic somewhat already existed (although through this PR I found it was buggy) with the generic "no thanks" option that was default added to the end of every preference set. The bug on master was that if option D is the generic-exclusive-no-thanks checkbox, if you click A B C then D it will correctly uncheck A-C, but if you then click any of A-C with D selected both will be checked, which is incorrect. This PR solves that issue.

In a separate PR before this which is merged, I added the ability to not automatically include the default exclusive box (through the `hideGenericDecline` flag), since that isn't part of OHA.

This PR adds the `exclusive` flag on a preference option which indicates it has the exclusive behavior outlined above. There can be any number of exclusive options. I had to abstract out some of the functions I wrote/updated so that the logic could be shared between public and partners (we needed the logic in parters in the piece where you can edit applications manually).

## Type of change


## How Can This Be Tested/Reviewed?

You will need to pull in the [Coliseum listing](https://github.com/housingbayarea/bloom/blob/0.3_alameda/backend/core/listing_data/coliseum.json), and check it out in public by going through the application process and checking/unchecking allllll the boxes. Then submit the application, head over to partners and edit it, and look at the checkboxes there for the same exclusive functionality on the last two options for OHA.

## Checklist:


** I would love to be testing this change but we don't really have a way to do that yet. Just another push to add a unit test suite to the public and partners site.
